cmd/gcp-controller-manager/node_annotator.go

```diff
@@ -271,17 +276,67 @@ func extractKubeLabels(instance *compute.Instance) (map[string]string, error) {
 		return make(map[string]string), nil
 	}
 
+	parsedLabels, err := parseLabels(*kubeLabels)
+	if err != nil {
+		return nil, fmt.Errorf("instance %q had %s", instance.SelfLink, err.Error())
+	}
+	return parsedLabels, nil
+}
+
+func extractLastAppliedLabels(node *core.Node) map[string]string {
+	if lastLabels, ok := node.ObjectMeta.Annotations[lastAppliedLabelsKey]; !ok || len(lastLabels) == 0 {
+		return make(map[string]string)
+	} else {
+		parsedLabels, err := parseLabels(lastLabels)
+		if err != nil {
+			klog.Errorf("failed to parse last applied labels annotation: %q, treat it as not set, err: %v", lastLabels, err)
+			return make(map[string]string)
+		}
+		return parsedLabels
+	}
+}
+
+func mergeManagedLabels(node *core.Node, desiredLabels map[string]string) error {
+	if node.ObjectMeta.Annotations == nil {
+		node.ObjectMeta.Annotations = make(map[string]string)
+	}
+	lastManagedLabels := extractLastAppliedLabels(node)
+	// Merge GCE managed labels by:
+	// 1. delete managed labels to be removed, which is present in last-applied-labels but missing in GCE label source
+	// 2. add/set labels from GCE label source to node, and last-applied-labels in annotation
+	for key := range lastManagedLabels {
+		if _, ok := desiredLabels[key]; !ok {
+			delete(node.ObjectMeta.Labels, key)
+		}
+	}
+	for key, value := range desiredLabels {
+		node.ObjectMeta.Labels[key] = value
+	}
+	node.ObjectMeta.Annotations[lastAppliedLabelsKey] = serializeLabels(desiredLabels)
```

Comment:
Looks like it will eventually be caught at Nodes.Update(node) but is there a limit to how many characters an annotation values can be?

Reply:
In documentation, it says annotation key is capped at 63 char + prefix, but value seems not capped.